RFC-0017-PyTorch-Operator-Versioning.md

```diff
@@ -0,0 +1,238 @@
+* It does not include the BC/FC for package/deploy itself. The Python-only operators are transparent to TS and Edge clients, with the TS compilation. 
+
+# Glossary
+
+* Backwards Compatibility (BC) — The ability to run programs from older versions of PyTorch
+* Version — A number that describes the format of the PyTorch program being read as well as providing partial information about which OpSet is required to run the program properly. (More precisely it counts the number of BC-breaking changes.) (See the dynamic versioning note (https://github.com/pytorch/pytorch/blob/6db8f7a70920f91418078fe09477eed0b0adefdb/caffe2/serialize/versions.h#L11).)
+* Forwards Compatibility (FC)* — The ability to run programs from future versions of PyTorch
+* Operator — A pair of a string (the operator’s “name” or “symbol”) and mathematical function, e.g. (div, /)
+* OpSet — A set of PyTorch operators (including upgraders)
+
+
+# Proposal
+
+We propose the operator versioning that works across eager, TorchScript, torch.package and mobile. It uses a version number + corresponding upgraders in torchscript to avoid breakage due to BC/FC breaking operator updates.
+
+* **Eager changes**
+    * `operator_versions.yaml` and `operator_upgraders.py` are added to register operator upgrades that are BC/FC breaking. 
+        * Note: this will not cover functional operators
+        * The default value is zero
+        * A version bump is also required for FC break only. It's is good for compatibility analysis: if the client is running old runtime, we don't deliver the new model with the un-compatible operator to avoid unexpected crash.
+    * **Use a single operator version number for all operators**
+        * This number may be shared by the deploy version, but separate from other file format versions
+    * **Older versions of operators registration are kept but they should only be matchable with special circumstances**
+    * **Newer version of the operator registry must specify an upgrader** that conforms to the older version of the operator schema. Its body is a TorchScript-able function that uses the newer version of operator to implement old semantics.
+        * One upgrader per historic signature. The registry specifies the symbol and the file formats those upgraders are applied to.
+    * [Improved BC testing] Tests that the old serialized version of the operator can still be loaded on the new runtime and run as expected need to be easy to add
+        * This seems straightforward for Torchscript and Edge testing but I’m not sure how it would work for deploy/package
+    * [Improved FC testing] Tests that the new version of the operator can still be loaded on old runtimes and run as expected need to be easy to add 
+        * This might require a new test job, which could be tricky to setup. We have no plans to support this.
```

Comment:
Maybe we can keep a record of the runtime with all ops every week. If FC requires 90 days, there will be ~12 runtimes in total

Reply:
Yes, it may be possible to keep runtime that is 1 version older (~90 days), but could be harder for CI tests setup with automatic updates.